kms/attestations/verify_attestation.py

```diff
@@ -0,0 +1,95 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This application verifies HSM attestations using certificate bundles
+obtained from Cloud HSM.
+
+For more information, visit https://cloud.google.com/kms/docs/attest-key.
+"""
+
+# [START verify_attestations]
+import argparse
+import gzip
+
+from cryptography import exceptions
+from cryptography import x509
+from cryptography.hazmat import backends
+from cryptography.hazmat.primitives.asymmetric import padding
+import pem
+
+
+def get_uncompressed_contents(filename):
+    """Gets the uncompressed contents of a file.
+
+    Args:
+        filename: The name of the file.
+
+    Returns:
+        The uncompressed contents of the file.
+    """
+    try:
+        with gzip.open(filename, 'rb') as f:
+            return f.read()
+    except IOError:
+        with open(filename, 'rb') as f:
+            return f.read()
+
+
+def verify(attestation_file, bundle_file):
+    """Verifies an attestation using a bundle of certificates.
+
+    Args:
+      attestation_file: The name of the attestation file.
+      bundle_file: The name of the bundle file containing the certificates
+        used to verify the attestation.
+
+    Returns:
+      True if at least one of the certificates in bundle_file can verify the
+      attestation statement and its signature.
+    """
+    attestation = get_uncompressed_contents(attestation_file)
+    statement = attestation[:-256]
+    signature = attestation[-256:]
+
+    for cert in pem.parse_file(bundle_file):
+        cert_obj = x509.load_pem_x509_certificate(
+            str(cert).encode('utf-8'), backends.default_backend())
+        try:
+            cert_obj.public_key().verify(
+                signature, statement, padding.PKCS1v15(),
+                cert_obj.signature_hash_algorithm)
+            return True
+        except exceptions.InvalidSignature:
+            continue
+    return False
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('attestation_file', help='The attestation filename.')
+    parser.add_argument(
+        'bundle_file',
+        help='The bundle filename containing one or more certificates.')
+
+    args = parser.parse_args()
+
+    if verify(args.attestation_file, args.bundle_file):
+        print('Signature verified!')
+    else:
+        print('Signature could not be verified.')
```

Comment:
There seems to be a lot more boilerplate code than needed for a sample. Could we do something like:

```
with gzip.open("my-attestation", 'rb') as f
   attestation = f.read()
```

and remove the helper and main functions?

Reply:
Like this?

Reply:
yup, looks great!